cmd/server.go

```diff
@@ -0,0 +1,125 @@
+package cmd
+
+import (
+	"context"
+	"net/http"
+	"os"
+	"os/signal"
+	"syscall"
+
+	"github.com/spf13/cobra"
+	"github.com/spf13/viper"
+	"github.com/usememos/memos/server"
+	_profile "github.com/usememos/memos/server/profile"
+
+	"fmt"
+)
+
+const (
+	greetingBanner = `
+███╗   ███╗███████╗███╗   ███╗ ██████╗ ███████╗
+████╗ ████║██╔════╝████╗ ████║██╔═══██╗██╔════╝
+██╔████╔██║█████╗  ██╔████╔██║██║   ██║███████╗
+██║╚██╔╝██║██╔══╝  ██║╚██╔╝██║██║   ██║╚════██║
+██║ ╚═╝ ██║███████╗██║ ╚═╝ ██║╚██████╔╝███████║
+╚═╝     ╚═╝╚══════╝╚═╝     ╚═╝ ╚═════╝ ╚══════╝
+`
+)
+
+var (
+	mode    string
+	port    int
+	data    string
+	feat    string
+	profile *_profile.Profile
+
+	rootCmd = &cobra.Command{
+		Use:   "memos",
+		Short: `An open-source, self-hosted memo hub with knowledge management and social networking.`,
+		Run: func(_cmd *cobra.Command, _args []string) {
+			ctx, cancel := context.WithCancel(context.Background())
+			s, err := server.NewServer(ctx, profile)
+			if err != nil {
+				cancel()
+				fmt.Printf("failed to create server, error: %+v\n", err)
+				return
+			}
+
+			c := make(chan os.Signal, 1)
+			// Trigger graceful shutdown on SIGINT or SIGTERM.
+			// The default signal sent by the `kill` command is SIGTERM,
+			// which is taken as the graceful shutdown signal for many systems, eg., Kubernetes, Gunicorn.
+			signal.Notify(c, os.Interrupt, syscall.SIGTERM)
+			go func() {
+				sig := <-c
+				fmt.Printf("%s received.\n", sig.String())
+				s.Shutdown(ctx)
+				cancel()
+			}()
+
+			println(greetingBanner)
+			fmt.Printf("Version %s has started at :%d\n", profile.Version, profile.Port)
+			if err := s.Start(ctx); err != nil {
+				if err != http.ErrServerClosed {
+					fmt.Printf("failed to start server, error: %+v\n", err)
+					cancel()
+				}
+			}
+
+			// Wait for CTRL-C.
+			<-ctx.Done()
+		},
+	}
+)
+
+func Execute() error {
+	return rootCmd.Execute()
+}
+
+func init() {
+	cobra.OnInitialize(initConfig)
+
+	rootCmd.PersistentFlags().StringVarP(&mode, "mode", "m", "demo", `mode of server, can be "prod" or "dev" or "demo"`)
+	rootCmd.PersistentFlags().IntVarP(&port, "port", "p", 8081, "port of server")
+	rootCmd.PersistentFlags().StringVarP(&data, "data", "d", "", "data directory")
+	rootCmd.PersistentFlags().StringVarP(&feat, "feat", "f", "", "feature flags, split by comma")
```

Comment:
Why do we need a flag to enable features? 

IMO, we should use it together with the `mode`. e.g. default enable these features in dev/demo mode but disable in prod mode.

Reply:
in the future, we can use feature args to enable/disable some advanced actions, which may hava break change, like store resources in local file system instead of db.

Reply:
for those people who want to still use resource in db, they can just run as before, but for others, they can use this feature manually

Reply:
> store resources in local file system instead of db.

As we already have system settings, maybe it will has a better UX to set the resource storage method in UI + system settings? 

![telegram-cloud-photo-size-5-6307292232071362379-y](https://user-images.githubusercontent.com/24653555/220367826-dc46fb90-52b9-4351-ab51-2a362c540f50.jpg)

Reply:
So actually... I am confused what exactly this PR brings.

Reply:
> > store resources in local file system instead of db.
> 
> As we already have system settings, maybe it will has a better UX to set the resource storage method in UI + system settings? 
> 
> ![telegram-cloud-photo-size-5-6307292232071362379-y](https://user-images.githubusercontent.com/24653555/220367826-dc46fb90-52b9-4351-ab51-2a362c540f50.jpg)
> 

looks good

Reply:
PTAL https://github.com/usememos/memos/pull/1140